src/recovery/recover_entity.rs

```diff
@@ -0,0 +1,1021 @@
+            continue;
+        };
+
+        let mut threshold_factor_instances = IndexSet::new();
+        let mut override_factor_instances = IndexSet::new();
+
+        for threshold_factor_hash in matrix_of_hashes.threshold_factors.iter() {
+            let Some(factor_instance) = map_hash_to_factor.get(threshold_factor_hash) else {
+                warn!(
+                    "Missing THRESHOLD factor instance for hash: {:?}",
+                    threshold_factor_hash
+                );
+                continue;
+            };
+            threshold_factor_instances.insert(factor_instance.clone());
+        }
+
+        for override_factor_hash in matrix_of_hashes.override_factors.iter() {
+            let Some(factor_instance) = map_hash_to_factor.get(override_factor_hash) else {
+                warn!(
+                    "Missing OVERRIDE factor instance for hash: {:?}",
+                    override_factor_hash
+                );
+                continue;
+            };
+            override_factor_instances.insert(factor_instance.clone());
+        }
+
+        if threshold_factor_instances.len() < matrix_of_hashes.threshold as usize {
+            warn!("Not enough threshold factors");
+            fail();
+            continue;
+        }
+
+        let sec = SecurifiedEntityControl::new(
+            MatrixOfFactorInstances::new(
+                threshold_factor_instances,
+                matrix_of_hashes.threshold,
+                override_factor_instances,
+            ),
+            on_chain_entity.access_controller,
+        );
+        let security_state = EntitySecurityState::Securified(sec);
+        let recovered_securified_entity =
+            E::new(format!("Recovered Securified: {:?}", a), a, security_state);
+        assert!(securified_entities.insert(recovered_securified_entity));
+    }
+
+    // L. Return the results, which is three sets: recovered_unsecurified, recovered_securified, unrecovered
+    Ok(EntityRecoveryOutcome::<E>::new(
+        unsecurified_entities,
+        securified_entities,
+        Vec::new(),
+    ))
+}
+
+pub async fn recover_accounts(
+    network_id: NetworkID,
+    factor_sources: impl IntoIterator<Item = HDFactorSource>,
+    key_derivation_interactors: Arc<dyn KeysDerivationInteractors>,
+    gateway: Arc<dyn GatewayReadonly>,
+) -> Result<EntityRecoveryOutcome<Account>> {
+    recover_entity(
+        network_id,
+        factor_sources,
+        key_derivation_interactors,
+        gateway,
+    )
+    .await
+}
+
+pub async fn recover_personas(
+    network_id: NetworkID,
+    factor_sources: impl IntoIterator<Item = HDFactorSource>,
+    key_derivation_interactors: Arc<dyn KeysDerivationInteractors>,
+    gateway: Arc<dyn GatewayReadonly>,
+) -> Result<EntityRecoveryOutcome<Persona>> {
+    recover_entity(
+        network_id,
+        factor_sources,
+        key_derivation_interactors,
+        gateway,
+    )
+    .await
+}
+
+#[cfg(test)]
+pub struct TestGateway {
```

Comment:
`TestGateway` needed to "mock" gateway, enabling writing of POC of reovery and tests.